File: controllers/preprocess.py
from flask import Blueprint, request, jsonify
from core.state import _state
from modules.preprocessing import full_preprocessing_pipeline

preprocess_bp = Blueprint('preprocess', __name__)

@preprocess_bp.route('/api/preprocess/single', methods=['POST'])
def preprocess_single():
    data = request.get_json()
    if not data or 'text' not in data:
        return jsonify({'success': False, 'error': 'Field "text" diperlukan'}), 400

    text         = data['text']
    verbose      = data.get('verbose', True)
    use_stemming = data.get('use_stemming', False)
    use_stopword = data.get('use_stopword', False)

    try:
        result = full_preprocessing_pipeline(
            text, _state['lexicon_merged'], verbose=verbose,
            use_stemming=use_stemming, use_stopword=use_stopword
        )
        return jsonify({'success': True, 'result': result})
    except Exception as e:
        return jsonify({'success': False, 'error': str(e)}), 500

@preprocess_bp.route('/api/preprocess/batch', methods=['POST'])
def preprocess_batch():
    if not _state['dataset']:
        return jsonify({'success': False,
                        'error': 'Belum ada dataset. Upload dataset terlebih dahulu.'}), 400

    data = request.get_json() or {}
    use_stemming = data.get('use_stemming', False)
    use_stopword = data.get('use_stopword', False)

    results = []
    preprocessed_corpus = []

    for item in _state['dataset']:
        processed = full_preprocessing_pipeline(
            item.get('text', ''),
            _state['lexicon_merged'],
            verbose=False,
            use_stemming=use_stemming,
            use_stopword=use_stopword
        )
        # Sertakan KEDUA format label untuk konsistensi:
        #   - 'label'  : string display untuk UI (e.g. "label: 1")
        #   - 'labels' : dict asli untuk SVM training (e.g. {'label': '1'})
        processed['label'] = item.get('label', '')
        processed['labels'] = item.get('labels', {})
        results.append(processed)
        preprocessed_corpus.append(processed['normalized'])

    # Simpan corpus yang sudah dipreprocess untuk digunakan TF-IDF
    _state['preprocessed_corpus'] = preprocessed_corpus
    _state['tfidf_result'] = None  # Reset TF-IDF karena corpus baru

    total = len(results)
    total_replacements = sum(r['stats']['replacements_made'] for r in results)

    return jsonify({
        'success': True,
        'total'  : total,
        'aggregate_stats': {
            'total_texts'       : total,
            'total_replacements': total_replacements,
            'avg_replacements'  : round(total_replacements / total, 2) if total else 0,
        },
        'results'          : results[:50],
        'corpus_ready'     : True,
    })

File: js/preprocess.js
// ── Preprocessing Single ──────────────────────────────────────────
const PP_EX = [
  'Dasar brengsek lo! gak tau diri bgt, pergi aja dari sini! https://bit.ly/abc #kebencian @user123 456',
  'Wkwk lucu bgt deh, tp gw gak setuju sama pendapat lo yg itu, gimana menurut org lain?',
  'Lo emang ga punya otak! <b>Blm</b> bisa mikir waras ya? https://spam.com @bot 99',
];
function ppEx(n) { document.getElementById('pp-inp').value=PP_EX[n-1]; ppRun(); }

async function ppRun() {
  const text = document.getElementById('pp-inp').value.trim();
  if (!text) return;
  const use_stopword = document.getElementById('pp-stopword')?.checked || false;
  const use_stemming = document.getElementById('pp-stemming')?.checked || false;
  const d = await api('/api/preprocess/single','POST',{text,verbose:true, use_stopword, use_stemming});
  if (!d.success) return;
  const r = d.result;
  document.getElementById('pp-raw').textContent  = r.raw;
  document.getElementById('pp-norm').textContent = r.normalized||'(kosong)';
  document.getElementById('pp-results').classList.remove('hidden');
  document.getElementById('pp-steps').innerHTML = (r.cleansing_steps||[]).map(([name,val],i)=>
    `<div class="step-item">
      <div class="step-num">${i+1}</div>
      <div class="step-body">
        <div class="step-name">${name}</div>
        <div class="step-val">${val||'<em class="text-mute">(kosong)</em>'}</div>
      </div>
    </div>`
  ).join('');
  const rc = r.replacements||[];
  document.getElementById('pp-rep-chip').textContent = rc.length+' penggantian';
  document.getElementById('pp-reps').innerHTML = rc.length>0
    ? '<div class="replace-list">'+rc.map(x=>`<span class="replace-tag"><span class="orig">${x.original}</span><span class="arr">→</span><span class="repl">${x.replaced||'∅'}</span></span>`).join('')+'</div>'
    : '<p class="text-sm text-mute" style="padding:8px 0;">Tidak ada penggantian.</p>';
  document.getElementById('pp-s-raw').textContent = r.stats.raw_length;
  document.getElementById('pp-s-cln').textContent = r.stats.cleansed_length;
  document.getElementById('pp-s-tok').textContent = r.stats.tokens_raw;
  document.getElementById('pp-s-rep').textContent = r.stats.replacements_made;
}

// ── Batch ─────────────────────────────────────────────────────────
async function checkBatchPre() {
  const ls = await api('/api/lexicon/status');
  const ts = await api('/api/tfidf/status');
  let html = '';
  if (ts.corpus_size===0)
    html += mkAlert('warn','Belum ada dataset. Upload dataset terlebih dahulu.');
  else
    html += mkAlert('success',`Dataset siap: ${ts.corpus_size} teks`);
  html += mkAlert('info',`Kamus gabungan: ${ls.merged_count} entri`);
  document.getElementById('batch-prereq').innerHTML = html;
}

async function runBatch() {
  const btn = document.getElementById('batch-btn');
  btn.disabled=true; btn.innerHTML='<span class="spinner"></span> Memproses…';
  const use_stopword = document.getElementById('b-stopword')?.checked || false;
  const use_stemming = document.getElementById('b-stemming')?.checked || false;
  const d = await api('/api/preprocess/batch','POST', {use_stopword, use_stemming});
  btn.disabled=false; btn.innerHTML='▶ Jalankan Batch Processing';
  if (!d.success) { alert(d.error); return; }
  document.getElementById('batch-results').classList.remove('hidden');
  document.getElementById('b-total').textContent = d.aggregate_stats.total_texts;
  document.getElementById('b-rep').textContent   = d.aggregate_stats.total_replacements;
  document.getElementById('b-avg').textContent   = d.aggregate_stats.avg_replacements;
  document.getElementById('batch-tbody').innerHTML = d.results.map((r,i)=>
    `<tr>
      <td class="text-mute">${i+1}</td>
      <td class="${r.label=='1'||r.label?.toLowerCase()=='hate'?'label-hate':'label-ok'}">${r.label}</td>
      <td title="${r.raw}">${r.raw}</td>
      <td title="${r.normalized}">${r.normalized}</td>
      <td style="color:var(--purple)">${r.stats.replacements_made}</td>
    </tr>`
  ).join('');
  loadDashStats();
}
